### PR TITLE
Feature: Enumerable

### PR DESCRIPTION
**Pull request by oxenprogrammer**
# Ruby's Enumerable Methods

Re-implementation of Ruby's in-built Enumerable for Microverse Study purpose.

## Built With

- Ruby
- VS Code
- Ubuntu
- Rubocop linter

## Methods

- my_each
- my_each_with_index
- my_select 
- my_all? 
- my_any?
- my_none?
- my_count
- my_map
- my_inject

### CODE REVIEW

File: lib/enumerable_method.rb [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                proc.call(element)
+              else
+                yield(element)
+              end
+    end
+    item
+  end
+
+  # my_inject
+  def my_inject(pa1 = nil, pa2 = nil)
```
Comment (kevcap) [inline lib/enumerable_method.rb:163]: On the #my_inject, we got only one error:
```
rspec ./spec/enumerables_spec.rb:404 # enumerables #my_inject raises a "LocalJumpError" when no block or argument is given

```

File: lib/enumerable_method.rb [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        count += 1 if yield item
+      else
+        count += 1
+      end
+    end
+    count
+  end
+
+  # my_map
+  def my_map(proc = nil)
```
Comment (kevcap) [inline lib/enumerable_method.rb:146]: On the #my_map we got just two errors:
```
rspec ./spec/enumerables_spec.rb:374 # enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
rspec ./spec/enumerables_spec.rb:389 # enumerables #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```

File: lib/enumerable_method.rb [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    else
+      my_each do |x|
+        return true unless x == false || x.nil?
+      end
+      false
+    end
+  end
+
+  # my_none?
+  def my_none?(parameter = nil)
```
Comment (kevcap) [inline lib/enumerable_method.rb:96]: On the #my_none? we got the following errors:
```
rspec ./spec/enumerables_spec.rb:323 # enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
rspec ./spec/enumerables_spec.rb:332 # enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
```

File: lib/enumerable_method.rb [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+          y += 1 unless x == false || x.nil?
+        end
+        true if y == size
+        y == size
+      end
+    end
+  end
+
+  # my_any?
+  def my_any?(parameter = nil)
```
Comment (kevcap) [inline lib/enumerable_method.rb:67]: On the #my_any? we got some errors as well:
```
rspec ./spec/enumerables_spec.rb:240 # enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
rspec ./spec/enumerables_spec.rb:250 # enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
rspec ./spec/enumerables_spec.rb:264 # enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

```

File: lib/enumerable_method.rb [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    elements = [] if is_a? Array
+    my_each do |item|
+      elements << item if yield item
+    end
+    elements
+  end
+
+  # my_all function
+  def my_all?(parameter = nil)
```
Comment (kevcap) [inline lib/enumerable_method.rb:39]: On the #my_all? method, we got a few errors:
```
rspec ./spec/enumerables_spec.rb:185 # enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
rspec ./spec/enumerables_spec.rb:195 # enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
rspec ./spec/enumerables_spec.rb:201 # enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

```

File: lib/enumerable_method.rb [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    my_each do |data|
+      yield data, i
+
+      i += 1
+    end
+    self
+  end
+
+  # my_select function
+  def my_select
```
Comment (kevcap) [inline lib/enumerable_method.rb:27]: On the #my_select method, we only got the following error:
```
rspec ./spec/enumerables_spec.rb:123 # enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value::range

```

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
## [REQUIRED CHANGES]
- [ ] We just have a few errors, which I typed down the reasons of the errors, and by checking the errors, you can know what's wrong and work to fix it 💪;

Tip to fix your errors: Remember to use a lot of types of inputs such as arrays, strings, words (w%[ ... ]), numbers, hashes, ranges (Range.new), blocks and even the array.clone method.

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In the real-world, you'll have a lot of change requests for your real clients, it's good to get used to it.</sub>

File: lib/enumerable_method.rb (hunk 6 above)
Comment (oxenprogrammer) [reply lib/enumerable_method.rb:27]: fixed for instance of Range

File: lib/enumerable_method.rb (hunk 2 above)
Comment (oxenprogrammer) [reply lib/enumerable_method.rb:146]: Fixed

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  good job! You were able to solve most of the errors and we're almost done, you are just required to do these changes, listed below:
 
## [REQUIRED CHANGES]
- [x] Now we only have *4 errors*, which I typed down the reasons of the errors, and by checking the errors, you can know what's wrong and work to fix it 💪;

1. The #my_any? when a pattern other than Regex or a Class is given, it *SHOULD* return true if any of the collection matches the pattern
2. The #my_none? when a pattern other than Regex or a Class is given, it  *SHOULD* return true only if none of the collection matches the pattern
3. The #my_none? when a pattern other than Regex or a Class is given, it *SHOULD* return false only if one of the collection matches the pattern
4. enumerables #multiply_els *SHOULD* accept an array as an argument and multiplies all the elements of the array together using #my_inject
```
Tip to fix your errors: Remember to use a lot of types of inputs such as arrays, strings, words (w%[ ... ]), numbers, hashes, ranges (Range.new), blocks and even the array.clone method.

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> <img  alt="gmail"  src="https://img.shields.io/badge/gmail-%20- green?logo=gmail&style=social"></a> [![Twitter Follow](https://img.shields.io/twitter/follow/kevinnnmn?style=social)](https://twitter.com/intent/follow?screen_name=kevinnnmn) [![](https://img.shields.io/badge/Slack-Kevin%20Alves-green?logo=slack&style=social)](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In the real-world, you'll have a lot of change requests for your real clients, it's good to get used to it.</sub>

**Review by mcrd25 — APPROVED** [review]
# changes approved :heavy_check_mark: 

Hi Team! :wave: 

You have done a good job! I see no need for further review! Thanks for making all the changes :+1: 

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

If you liked my review, consider upskilling me on LinkedIn. And please do connect. [My LinkedIn](https://www.linkedin.com/in/mayadouglas/)


*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**